### PR TITLE
Visual Recognition : added a new Delete All Classifiers feature + 0.4.4 merges.

### PR DESCRIPTION
**Pull request by ylecleach**
5 days passed to fix all codacy issues. Code is in better shape, definitively. Thank you Codacy !

### CODE REVIEW

File: services/visual_recognition/v3.js [hunk 1]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+          text:'Delete All : call to listClassifiers failed'});
+        node.error(err, msg);   
+      } else {
+        // Array to hold async tasks
+        var asyncTasks=[], nbTodelete=0, nbdeleted=0;
+        nbTodelete = body.classifiers.length;
+        body.classifiers.forEach(function (aClassifier) {
+          asyncTasks.push(function (cb) {
+            var parms = {};
+            parms.classifier_id=aClassifier.classifier_id;
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'classifier_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l233)

File: services/visual_recognition/v3.js [hunk 2]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+      node.status({fill:'red', shape:'ring', 
+                   text:'call to watson visual recognition v3 service failed'}); 
+      msg.result = {};
+      msg.result['error_id']= body.images[0].error.error_id;
+      msg.result['error']= body.images[0].error.description;
+      msg.payload='see msg.result.error';
+      node.send(msg); 
+    } else {
+      if (feature === 'deleteClassifier') {
+        msg.result = 'Successfully deleted classifier_id: '+ msg.params.classifier_id ;
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'classifier_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l127)

File: services/visual_recognition/v3.js (hunk 1 above)
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'classifier_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l233)

File: services/visual_recognition/v3.js [hunk 3]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+    // If it is present the newly provided user entered key 
+    // takes precedence over the existing one. 
+    node.apikey = sAPIKey || node.credentials.apikey;
+    if (!node.apikey) {
+      node.status({fill:'red', shape:'ring', text:'missing credentials'});          
+      var message ='Missing Watson Visual Recognition API service credentials'; 
+      node.error(message, msg);
+      return false;
+    }
+    node.service = watson.visual_recognition({
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'visual_recognition' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l92)

File: services/visual_recognition/v3.js [hunk 4]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+    if (imageCheck(msg.payload)) {
+      temp.open({suffix: '.' + fileType(msg.payload).ext}, function (err, info) {
+        if (err) {
+          this.status({fill:'red', shape:'ring', text:'unable to open image stream'});          
+          node.error('Node has been unable to open the image stream', msg);
+          return cb();
+        }  
+        stream_buffer(info.path, msg.payload, function () {
+          params['images_file'] = fs.createReadStream(info.path);
+          if (msg.params != null && msg.params.classifier_ids != null) {
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'classifier_ids' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l147)

File: services/visual_recognition/v3.js [hunk 5]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
-          node.status({});
-        }
-      }; // actionComplete    
-
-      var actionComplete2 = function(err, body, other) {
+                       text:'unable to open image stream'});          
+          node.error('Node has been unable to open the image stream', msg);
+          return callback('open error on '+key);
+        }  
+        stream_buffer(info.path, msg.params[key], function () {
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'stream_buffer' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l188)

File: services/visual_recognition/v3.js [hunk 6]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
-        return;
+  function prepareParamsCommon(params, node, msg, cb) {
+    if (imageCheck(msg.payload)) {
+      temp.open({suffix: '.' + fileType(msg.payload).ext}, function (err, info) {
+        if (err) {
+          this.status({fill:'red', shape:'ring', text:'unable to open image stream'});          
+          node.error('Node has been unable to open the image stream', msg);
+          return cb();
+        }  
+        stream_buffer(info.path, msg.payload, function () {
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'stream_buffer' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l145)

File: services/visual_recognition/v3.js [hunk 7]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+    // takes precedence over the existing one. 
+    node.apikey = sAPIKey || node.credentials.apikey;
+    if (!node.apikey) {
+      node.status({fill:'red', shape:'ring', text:'missing credentials'});          
+      var message ='Missing Watson Visual Recognition API service credentials'; 
+      node.error(message, msg);
+      return false;
+    }
+    node.service = watson.visual_recognition({
+      api_key: node.apikey,
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'api_key' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l93)

File: services/visual_recognition/v3.js [hunk 8]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+              if (err) {
+                node.error(err, msg);
+                //console.log('Error with the removal of classifier_id ' +
+                  //parms.classifier_id +' : ' + err);
+                return cb('error');
+              }
+              //console.log('Classifier ID '+ aClassifier.classifier_id + 
+              //  ' deleted successfully.');
+              nbdeleted++;
+              cb(null,parms.classifier_id);
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'classifier_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l244)

File: services/visual_recognition/v3.js [hunk 9]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
-          return;
-        }
+      node.error(err);
+      return;
+    } else if (err == null && body != null && body.images != null && 
+      body.images[0].error) {
+      node.status({fill:'red', shape:'ring', 
+                   text:'call to watson visual recognition v3 service failed'}); 
+      msg.result = {};
+      msg.result['error_id']= body.images[0].error.error_id;
```
Comment (chughts) [inline services/visual_recognition/v3.js]: Codacy Issue found: [Identifier 'error_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737047177/issues/source?bid=3330694&fileBranchId=3352492#l121)

File: services/visual_recognition/v3.js [hunk 10]
```diff
@@ -14,46 +14,21 @@
...
-  var async = require('async');
+  var cfenv = require('cfenv'),
+    watson = require('watson-developer-cloud'),
+    imageType = require('image-type'),
+    url = require('url'),
+    temp = require('temp'),
+    fileType = require('file-type'),
+    fs = require('fs'),
+    async = require('async'),
+    apikey = null, sAPIKey=null, service=null;
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['apikey' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l26)

File: services/visual_recognition/v3.js [hunk 11]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+        });
+      }
+    }); 
+  }  // delete all func 
+
+  function executeService(feature, params, node, msg) {
+    switch(feature) {
+    case 'classifyImage':
+      prepareParamsCommon(params, node, msg, function () {
+        node.service.classify(params, function(err, body, other) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['other' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l268)

File: services/visual_recognition/v3.js [hunk 12]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+      break;
+    case 'retrieveClassifierDetails':
+      params['classifier_id']=msg.params['classifier_id'];
+      node.service.getClassifier(params, function(err, body, other) {
+        processResponse(err,body,feature,node,msg);
+      });
+      break;
+    case 'deleteClassifier':
+      params['classifier_id']=msg.params['classifier_id'];
+      node.service.deleteClassifier(params, function(err, body, other) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['other' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l312)

File: services/visual_recognition/v3.js [hunk 13]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+      });
+      break;
+    case 'retrieveClassifiersList':
+      node.service.listClassifiers(params, function(err, body, other) {
+        processResponse(err,body,feature,node,msg);
+      });
+      break;
+    case 'retrieveClassifierDetails':
+      params['classifier_id']=msg.params['classifier_id'];
+      node.service.getClassifier(params, function(err, body, other) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['other' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l306)

File: services/visual_recognition/v3.js [hunk 14]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+  function prepareParamsCreateClassifier (params, node, msg, cb) {
+    var listParams = {}, asyncTasks = [] , key=null;
+    for (key in msg.params) {
+      if (key.indexOf('_examples')>=0) {
+        addTask(asyncTasks, msg, key, listParams, node);
+      } else if (key==='name') {
+        listParams[key]=msg.params[key];
+      }
+    } // for
+    async.parallel(asyncTasks, function(error, results){
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['results' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l205)

File: services/visual_recognition/v3.js [hunk 15]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+        node.error(err, msg);   
+      } else {
+        // Array to hold async tasks
+        var asyncTasks=[], nbTodelete=0, nbdeleted=0;
+        nbTodelete = body.classifiers.length;
+        body.classifiers.forEach(function (aClassifier) {
+          asyncTasks.push(function (cb) {
+            var parms = {};
+            parms.classifier_id=aClassifier.classifier_id;
+            node.service.deleteClassifier(parms, function(err, body) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['body' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l234)

File: services/visual_recognition/v3.js [hunk 16]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+    case 'detectFaces':
+      prepareParamsCommon(params, node, msg, function () {
+        node.service.detectFaces(params, function(err, body, other) {
+          processResponse(err,body,feature,node,msg);
+        });
+      });
+      break;
+    case 'recognizeText':
+      prepareParamsCommon(params, node, msg, function () {
+        node.service.recognizeText(params, function(err, body, other) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['other' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l282)

File: services/visual_recognition/v3.js [hunk 17]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+    case 'classifyImage':
+      prepareParamsCommon(params, node, msg, function () {
+        node.service.classify(params, function(err, body, other) {
+          processResponse(err,body,feature,node,msg);
+        });
+      });
+      break;
+    case 'detectFaces':
+      prepareParamsCommon(params, node, msg, function () {
+        node.service.detectFaces(params, function(err, body, other) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['other' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l275)

File: services/visual_recognition/v3.js [hunk 18]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
+    switch(feature) {
+    case 'createClassifier':
+      prepareParamsCreateClassifier(params, node, msg, function () {
+        node.service.createClassifier(params, function(err, body, other) {
+          processResponse(err,body,feature,node,msg);
+        });
+      });
+      break;
+    case 'retrieveClassifiersList':
+      node.service.listClassifiers(params, function(err, body, other) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['other' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l300)

File: services/visual_recognition/v3.js [hunk 19]
```diff
@@ -62,301 +37,331 @@ module.exports = function (RED) {
...
-                      listParams[prop]=fs.createReadStream(info.path);
-                      callback(null, prop);
-                    });
-                  }); // temp.open
-              }); // asyncTasks.push
+  function executeUtilService(feature, params, node, msg) {
+    switch(feature) {
+    case 'createClassifier':
+      prepareParamsCreateClassifier(params, node, msg, function () {
+        node.service.createClassifier(params, function(err, body, other) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js]: Codacy Issue found: ['other' is defined but never used.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737047197/issues/source?bid=3225115&fileBranchId=3352493#l294)

File: services/visual_recognition/v3.js [hunk 20]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+        stream_buffer(info.path, msg.params[k], function () {
+          listParams[k] = fs.createReadStream(info.path);
+          callback(null, k);
+        });
+      });
+    });
+  }
+
+  function prepareParamsCreateClassifier (params, node, msg, cb) {
+    var listParams = {}, asyncTasks = [] , k = null;
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:198]: Codacy Issue found: [Expected blank line after variable declarations.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l198)

File: services/visual_recognition/v3.js [hunk 21]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
-      // takes precedence over the existing one. 
-      apikey = sAPIKey || this.credentials.apikey;
-      this.status({}); 
-
-      if (!apikey) {
-        this.status({fill:'red', shape:'ring', text:'missing credentials'});          
-        message ='Missing Watson Visual Recognition API service credentials'; 
-        node.error(message, msg);
-        return;
+  function prepareParamsCommon(params, node, msg, cb) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:138]: Codacy Issue found: [Function 'prepareParamsCommon' has a complexity of 6.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l138)

File: services/visual_recognition/v3.js [hunk 22]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+  function verifyPayload(node, msg) {
+    if (!msg.payload) {
+      this.status({fill:'red', shape:'ring', text:'missing payload'}); 
+      node.error('Missing property: msg.payload', msg);
+      return false;
+    }
+    return true;
+  }
+
+  function verifyInputs(feature, node, msg) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:68]: Codacy Issue found: [Function 'verifyInputs' has a complexity of 6.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l68)

File: services/visual_recognition/v3.js [hunk 23]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+      msg.result['error'] = body.images[0].error.description;
+      msg.payload = 'see msg.result.error';
+      node.send(msg); 
+    } else {
+      if (feature === 'deleteClassifier') {
+        msg.result = 'Successfully deleted classifier_id: ' + msg.params.classifier_id ;
+      } else {
+        msg.result = body;
       }
+      msg.payload = 'see msg.result'; // to remove any Buffer that could remains
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:132]: Codacy Issue found: [Unexpected comment inline with code.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l132)

File: services/visual_recognition/v3.js [hunk 24]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+  function performDeleteAllClassifiers(params, node, msg) {
+    node.service.listClassifiers(params, function(err, body) {
+      node.status({});
+      if (err) {
+        node.status({fill:'red', shape:'ring', 
+          text:'Delete All : call to listClassifiers failed'});
+        node.error(err, msg);   
+      } else {
+        // Array to hold async tasks
+        var asyncTasks = [], nbTodelete = 0, nbdeleted = 0;
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:228]: Codacy Issue found: [Expected blank line after variable declarations.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l228)

File: services/visual_recognition/v3.js [hunk 25]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
-                    }  
-                    stream_buffer(info.path, msg.params[prop], function () {
-                      listParams[prop]=fs.createReadStream(info.path);
-                      callback(null, prop);
-                    });
-                  }); // temp.open
-              }); // asyncTasks.push
+  // This is the Watson Visual Recognition V3 Node
+  function WatsonVisualRecognitionV3Node (config) {
+    var node = this, b = false, feature = config['image-feature'];
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:330]: Codacy Issue found: [Expected blank line after variable declarations.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l330)

File: services/visual_recognition/v3.js [hunk 26]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+        node.service.recognizeText(params, function(err, body) {
+          processResponse(err,body,feature,node,msg);
+        });
+      });
+      break;
+    }
+  }
 
-      var actionCompleteDeleteClassifier = function(err, body, other) {
+  function executeUtilService(feature, params, node, msg) {
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:287]: Codacy Issue found: [Function 'executeUtilService' has a complexity of 6.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l287)

File: services/visual_recognition/v3.js [hunk 27]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
-          node.error('Error code : ' +  err.code);
-          return;
-        }
-        else if (body.images[0].error)
-        {
-          var err_desc = body.images[0].error.description;
-          var err_id = body.images[0].error.error_id;
+  function addTask (asyncTasks, msg, k, listParams, node) {
+    asyncTasks.push(function (callback) {
+      var buffer = msg.params[k];
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:181]: Codacy Issue found: [Expected blank line after variable declarations.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l181)

File: services/visual_recognition/v3.js [hunk 28]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
-  // the cleanup removes the temp storage, and I am not sure whether 
-  // it should be called here or after alchemy returns and passed
-  // control back to cbdone.
+  function verifyServiceCredentials(node, msg) {
+    // If it is present the newly provided user entered key 
+    // takes precedence over the existing one. 
+    node.apikey = sAPIKey || node.credentials.apikey;
+    if (!node.apikey) {
+      node.status({fill:'red', shape:'ring', text:'missing credentials'});          
+      var message ='Missing Watson Visual Recognition API service credentials';
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:88]: Codacy Issue found: [Infix operators must be spaced.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l88)

File: services/visual_recognition/v3.js [hunk 29]
```diff
@@ -14,46 +14,21 @@
...
-  var async = require('async');
+  var cfenv = require('cfenv'),
+    watson = require('watson-developer-cloud'),
+    imageType = require('image-type'),
+    url = require('url'),
+    temp = require('temp'),
+    fileType = require('file-type'),
+    fs = require('fs'),
+    async = require('async'),
+    sAPIKey=null, service=null;
```
Comment (yrezgui) [inline services/visual_recognition/v3.js:26]: Codacy Issue found: [Infix operators must be spaced.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2737057949/issues/source?bid=3225115&fileBranchId=3352493#l26)

File: services/visual_recognition/v3.js [hunk 30]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+    if (imageCheck(msg.payload)) {
+      temp.open({suffix: '.' + fileType(msg.payload).ext}, function (err, info) {
+        if (err) {
+          this.status({fill:'red', shape:'ring', text:'unable to open image stream'});          
+          node.error('Node has been unable to open the image stream', msg);
+          return cb();
+        }  
+        stream_buffer(info.path, msg.payload, function () {
+          params['images_file'] = fs.createReadStream(info.path);
+          if (msg.params != null && msg.params.classifier_ids != null) {
```
Comment (chughts) [inline services/visual_recognition/v3.js:148]: Codacy Issue found: [Identifier 'classifier_ids' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l148)

File: services/visual_recognition/v3.js [hunk 31]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
-          msg.result = {};
-          msg.result['all'] = body;
-          msg.payload='see msg.result'; // to remove any Buffer that could remains
-          node.send(msg); 
-          node.status({});
+                       text:'unable to open image stream'});          
+          node.error('Node has been unable to open the image stream', msg);
+          return callback('open error on ' + k);
+        }  
+        stream_buffer(info.path, msg.params[k], function () {
```
Comment (chughts) [inline services/visual_recognition/v3.js:189]: Codacy Issue found: [Identifier 'stream_buffer' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l189)

File: services/visual_recognition/v3.js [hunk 32]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+      node.status({fill:'red', shape:'ring', text:'missing credentials'});          
+      var message ='Missing Watson Visual Recognition API service credentials';
 
-  function performAction(params, feature, cbdone) {
-    var visualRecognition = watson.visual_recognition({
-      api_key: apikey,
+      node.error(message, msg);
+      return false;
+    }
+    node.service = watson.visual_recognition({
```
Comment (chughts) [inline services/visual_recognition/v3.js:93]: Codacy Issue found: [Identifier 'visual_recognition' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l93)

File: services/visual_recognition/v3.js [hunk 33]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
-        return;
+  function prepareParamsCommon(params, node, msg, cb) {
+    if (imageCheck(msg.payload)) {
+      temp.open({suffix: '.' + fileType(msg.payload).ext}, function (err, info) {
+        if (err) {
+          this.status({fill:'red', shape:'ring', text:'unable to open image stream'});          
+          node.error('Node has been unable to open the image stream', msg);
+          return cb();
+        }  
+        stream_buffer(info.path, msg.payload, function () {
```
Comment (chughts) [inline services/visual_recognition/v3.js:146]: Codacy Issue found: [Identifier 'stream_buffer' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l146)

File: services/visual_recognition/v3.js [hunk 34]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+      node.status({fill:'red', shape:'ring', 
+                   text:'call to watson visual recognition v3 service failed'}); 
+      msg.result = {};
+      msg.result['error_id'] = body.images[0].error.error_id;
+      msg.result['error'] = body.images[0].error.description;
+      msg.payload = 'see msg.result.error';
+      node.send(msg); 
+    } else {
+      if (feature === 'deleteClassifier') {
+        msg.result = 'Successfully deleted classifier_id: ' + msg.params.classifier_id ;
```
Comment (chughts) [inline services/visual_recognition/v3.js:128]: Codacy Issue found: [Identifier 'classifier_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l128)

File: services/visual_recognition/v3.js [hunk 35]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+        node.error(err, msg);   
+      } else {
+        // Array to hold async tasks
+        var asyncTasks = [], nbTodelete = 0, nbdeleted = 0;
+        nbTodelete = body.classifiers.length;
+        body.classifiers.forEach(function (aClassifier) {
+          asyncTasks.push(function (cb) {
+            var parms = {};
+
+            parms.classifier_id = aClassifier.classifier_id;
```
Comment (chughts) [inline services/visual_recognition/v3.js:234]: Codacy Issue found: [Identifier 'classifier_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l234)

File: services/visual_recognition/v3.js [hunk 36]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
-          return;
-        }
+      node.error(err);
+      return;
+    } else if (err == null && body != null && body.images != null && 
+      body.images[0].error) {
+      node.status({fill:'red', shape:'ring', 
+                   text:'call to watson visual recognition v3 service failed'}); 
+      msg.result = {};
+      msg.result['error_id'] = body.images[0].error.error_id;
```
Comment (chughts) [inline services/visual_recognition/v3.js:122]: Codacy Issue found: [Identifier 'error_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l122)

File: services/visual_recognition/v3.js [hunk 37]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+      var message ='Missing Watson Visual Recognition API service credentials';
 
-  function performAction(params, feature, cbdone) {
-    var visualRecognition = watson.visual_recognition({
-      api_key: apikey,
+      node.error(message, msg);
+      return false;
+    }
+    node.service = watson.visual_recognition({
+      api_key: node.apikey,
```
Comment (chughts) [inline services/visual_recognition/v3.js:94]: Codacy Issue found: [Identifier 'api_key' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l94)

File: services/visual_recognition/v3.js [hunk 38]
```diff
@@ -62,301 +37,329 @@ module.exports = function (RED) {
...
+            var parms = {};
+
+            parms.classifier_id = aClassifier.classifier_id;
+            node.service.deleteClassifier(parms, function(err) {
+              if (err) {
+                node.error(err, msg);
+                return cb('error');
+              }
+              nbdeleted++;
+              cb(null,parms.classifier_id);
```
Comment (chughts) [inline services/visual_recognition/v3.js:241]: Codacy Issue found: [Identifier 'classifier_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l241)

File: services/visual_recognition/v3.js (hunk 35 above)
Comment (chughts) [inline services/visual_recognition/v3.js:234]: Codacy Issue found: [Identifier 'classifier_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2737057989/issues/source?bid=3330694&fileBranchId=3352492#l234)

**Pull request merged**